tests/unit_tests/test_errors.py

```diff
@@ -0,0 +1,179 @@
+    def test_unknown_library(self):
+        """Test unknown library errors."""
+
+        err = FabUnknownLibraryError("mylib")
+        assert str(err) == "unknown library mylib"
+
+
+class TestToolErrors:
+    """Test the FabToolError hierarchy."""
+
+    def test_tool_string(self):
+        """Test the base FabToolError class."""
+
+        err = FabToolError("cc", "compiler message")
+        assert str(err) == "[cc] compiler message"
+
+        # Mock defines an internal name property, so special handling
+        # is required to reset the value to support testing
+        tool = Mock(_name="tool cc")
+        del tool.name
+        err = FabToolError(tool, "compiler message")
+        assert str(err) == "[tool cc] compiler message"
+
+        category = Mock()
+        del category._name
+        type(category).name = PropertyMock(return_value="category cc")
+        err = FabToolError(category, "compiler message")
+        assert str(err) == "[category cc] compiler message"
+
+    def test_mismatch(self):
+        """Test tool type mismatch class."""
+
+        err = FabToolMismatch("cc", "CCompiler", "Ar")
+        assert str(err) == "[cc] got type CCompiler instead of Ar"
+
+    def test_invalid_version(self):
+        """Test invalid version class."""
+
+        err = FabToolInvalidVersion("cc", "abc")
+        assert str(err) == "[cc] invalid version 'abc'"
+
+        err = FabToolInvalidVersion("cc", "abc", "VV.NN")
+        assert str(err) == "[cc] invalid version 'abc' should be 'VV.NN'"
+
+    def test_psyclone_api(self):
+        """Test PSyclone API  class."""
+
+        err = FabToolPsycloneAPI(None, "alg_file")
+        assert str(err) == "[psyclone] called without API but not with alg_file"
+
+        err = FabToolPsycloneAPI("nemo", "alg_file")
+        assert str(err) == "[psyclone] called with nemo API but not with alg_file"
+
+        err = FabToolPsycloneAPI("nemo", "alg_file", present=True)
+        assert str(err) == "[psyclone] called with nemo API and with alg_file"
+
+    def test_not_available(self):
+        """Test tool not available class."""
+
+        err = FabToolNotAvailable("psyclone")
+        assert str(err) == "[psyclone] not available"
+
+        err = FabToolNotAvailable("gfortran", "GCC")
+        assert str(err) == "[gfortran] not available in suite GCC"
+
+    def test_invalid_setting(self):
+        """Test invalid setting class."""
+
+        err = FabToolInvalidSetting("category", "compiler")
+        assert str(err) == "[compiler] invalid category"
+
+        err = FabToolInvalidSetting("category", "compiler", "nosuch")
+        assert str(err) == "[compiler] invalid category nosuch"
+
+
+class TestCommandErrors:
+    """Test various command errors."""
+
+    def test_command(self):
+        """Test FabCommandError in various configurations."""
+
+        err = FabCommandError(["ls", "-l", "/nosuch"], 1, b"", b"ls: cannot", "/")
+        assert str(err) == "command 'ls -l /nosuch' returned 1"
```

Comment:
Would it be easier to read if the sentence was restructured to put the variable length element last. e.g. "Return code 1 from command: ls -l /nosuch".

This way the eye knows where to go to for the error code and the start of the command.